src/vdaf/mastic.rs

```diff
@@ -379,32 +382,418 @@ where
+            szk_query_share_opt,
+        })
+    }
+}
+
+/// Mastic prepare message.
+///
+/// Result of preprocessing the broadcast messages of both aggregators during the
+/// preparation phase.
+pub type MasticPrepareMessage<const SEED_SIZE: usize> = SzkJointShare<SEED_SIZE>;
+
+impl<F: FieldElement, const SEED_SIZE: usize> ParameterizedDecode<MasticPrepareState<F, SEED_SIZE>>
+    for MasticPrepareMessage<SEED_SIZE>
+{
+    fn decode_with_param(
+        prep_state: &MasticPrepareState<F, SEED_SIZE>,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match prep_state.szk_query_state {
+            Some(_) => SzkJointShare::<SEED_SIZE>::decode_with_param(&true, bytes),
+            None => SzkJointShare::<SEED_SIZE>::decode_with_param(&false, bytes),
+        }
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Aggregator<SEED_SIZE, NONCE_SIZE> for Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    type PrepareState = MasticPrepareState<T::Field, SEED_SIZE>;
+    type PrepareShare = MasticPrepareShare<T::Field, SEED_SIZE>;
+    type PrepareMessage = MasticPrepareMessage<SEED_SIZE>;
+
+    fn prepare_init(
+        &self,
+        verify_key: &[u8; SEED_SIZE],
+        agg_id: usize,
+        agg_param: &MasticAggregationParam,
+        nonce: &[u8; NONCE_SIZE],
+        public_share: &MasticPublicShare<VidpfWeight<T::Field>>,
+        input_share: &MasticInputShare<T::Field, SEED_SIZE>,
+    ) -> Result<
+        (
+            MasticPrepareState<T::Field, SEED_SIZE>,
+            MasticPrepareShare<T::Field, SEED_SIZE>,
+        ),
+        VdafError,
+    > {
+        let id = match agg_id {
+            0 => Ok(VidpfServerId::S0),
+            1 => Ok(VidpfServerId::S1),
+            _ => Err(VdafError::Uncategorized(
+                "Invalid aggregator ID".to_string(),
+            )),
+        }?;
+        let mut eval_proof = P::init(
+            verify_key,
+            &self.domain_separation_tag(DST_PATH_CHECK_BATCH),
+        );
+        let mut output_shares = Vec::<T::Field>::with_capacity(
+            self.vidpf.weight_parameter * agg_param.level_and_prefixes.prefixes().len(),
+        );
+        let mut cache_tree = BinaryTree::<VidpfEvalCache<VidpfWeight<T::Field>>>::default();
+        let cache = VidpfEvalCache::<VidpfWeight<T::Field>>::init_from_key(
+            id,
+            &input_share.vidpf_key,
+            &self.vidpf.weight_parameter,
+        );
+        cache_tree
+            .insert(Path::empty(), cache)
+            .expect("Should alwys be able to insert into empty tree at root");
+        for prefix in agg_param.level_and_prefixes.prefixes() {
+            let mut value_share = self.vidpf.eval_with_cache(
+                id,
+                &input_share.vidpf_key,
+                public_share,
+                prefix,
+                &mut cache_tree,
+                nonce,
+            )?;
+            eval_proof.update(&value_share.proof);
```

Comment:
The eval_proof includes the node proofs, but it also needs the counter and path consistency checks. At this point, let's not try to add this to the current PR. Instead, please note this as a TODO in #947. Also, we may want to add this feature only after we've addressed https://github.com/jimouris/draft-mouris-cfrg-mastic/issues/92, as this should make the implementation easier.